test/Test.hs

```diff
@@ -22,8 +44,413 @@ monadReader2 = assertEqual "should be equal" [(5, 0)] $
       y <- ask
       return (x,y)
 
+monadReader3 :: Assertion
+monadReader3 = assertEqual "should be equal" [5,3] $
+  runReader (observeAllT (plus5 `mplus` mzero `mplus` plus3)) (0 :: Int)
+  where
+    plus5 = local (5+) ask
+    plus3 = local (3+) ask
+
+nats, odds, oddsOrTwo,
+  oddsOrTwoUnfair, oddsOrTwoFair,
+  odds5down :: Monad m => LogicT m Integer
+
+#if MIN_VERSION_base(4,8,0)
+nats = pure 0 `mplus` ((1 +) <$> nats)
+#else
+nats = return 0 `mplus` liftM (1 +) nats
+#endif
+
+odds = return 1 `mplus` liftM (2+) odds
+
+oddsOrTwoUnfair = odds `mplus` return 2
+oddsOrTwoFair   = odds `interleave` return 2
+
+oddsOrTwo = do x <- oddsOrTwoFair
+               if even x then once (return x) else mzero
+
+odds5down = return 5 `mplus` mempty `mplus` mempty `mplus` return 3 `mplus` return 1
+
+yieldWords :: [String] -> LogicT m String
+yieldWords = go
+  where go [] = mzero
+        go (w:ws) = return w `mplus` go ws
+
+
 main :: IO ()
-main = defaultMain $ testGroup "All"
+main = defaultMain $
+  localOption (mkTimeout 1000000) $  -- 1 second
+  testGroup "All"
+  [ testGroup "Monad Reader + env"
     [ testCase "Monad Reader 1" monadReader1
     , testCase "Monad Reader 2" monadReader2
+    , testCase "Monad Reader 3" monadReader3
+    ]
+
+  , testGroup "Various monads"
+    [
+      -- nats will generate an infinite number of results; demonstrate
+      -- various ways of observing them via Logic/LogicT
+      testCase "runIdentity all"  $ [0..4] @=? (take 5 $ runIdentity $ observeAllT nats)
+    , testCase "runIdentity many" $ [0..4] @=? (runIdentity $ observeManyT 5 nats)
+    , testCase "observeAll"       $ [0..4] @=? (take 5 $ observeAll nats)
+    , testCase "observeMany"      $ [0..4] @=? (observeMany 5 nats)
+
+    -- Ensure LogicT can be run over other base monads other than
+    -- List.  Some are productive (Reader) and some are non-productive
+    -- (Except) in the observeAll case.
+
+    , testCase "runReader" $ [0..4] @=? (take 5 $ runReader (observeAllT nats) "!")
+
+    , testCase "manyT runExceptT" $ [0..4] @=?
+      (either (const []) id $ runExcept (observeManyT 5 nats))
```

Comment:
```suggestion
      (either (const []) id $ observeManyT 5 nats)
```
How about this? It would allow to remove `import Control.Monad.Except` without CPP.

Reply:
I was hoping to explicitly test this: https://github.com/Bodigrim/logict/blob/master/Control/Monad/Logic.hs#L101

Reply:
Well, you can change that example to use `Either a` monad instead of `Except`.

Reply:
Good call, I've updated this.